### PR TITLE
Dev flask poc auth

### LINKED ISSUE #2564: API: Flask - Authentication
This issue is related to #2413 

We must design the best way to secure API access using openapi specification. The main idea is to implement an endpoint to get a auth token. This endpoint (GET /token) requires user/pass authentication and returns a valid token as a response, which has a limited life time (set up in API settings).

Any other endpoint should require a valid token in order to serve a response.

Some documentation:
* https://connexion.readthedocs.io/en/latest/security.html

Tasks:
- [x] Add GET /token to openapi spec
- [x] Implement token generation
- [x] Implement token life time limitation
- [x] ~~Get both http and https working, especially the last one~~

### PR DESCRIPTION
**Pull request by crd1985**
Hi team,

this PR closes #2564 . It implements the base to add, check and login users. 

It would still remain linking this new functionality to the new openapi specification, which is pending of completion.

This is the output of the unit tests:
root@23e7fc9958d4:/# /var/ossec/framework/python/bin/pytest /var/ossec/framework/python/lib/python3.7/site-packages/api-3.9.0-py3.7.egg/api/test/test_authentication.py 
================================================ test session starts =================================================
platform linux -- Python 3.7.2, pytest-4.3.0, py-1.8.0, pluggy-0.9.0
rootdir: /var/ossec/framework/python/lib/python3.7/site-packages/api-3.9.0-py3.7.egg/api/test, inifile:
collected 4 items                                                                                                    

var/ossec/framework/python/lib/python3.7/site-packages/api-3.9.0-py3.7.egg/api/test/test_authentication.py ... [ 75%]
.                                                                                                              [100%]

================================================== warnings summary ==================================================
test_authentication.py::test_database_init
  /var/ossec/framework/python/lib/python3.7/site-packages/jose/jwt.py:5: DeprecationWarning: Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated, and in 3.8 it will stop working
    from collections import Mapping

test_authentication.py::test_database_init
  /var/ossec/framework/python/lib/python3.7/site-packages/jose/jws.py:6: DeprecationWarning: Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated, and in 3.8 it will stop working
    from collections import Mapping, Iterable

test_authentication.py::test_database_init
test_authentication.py::test_database_init
  /var/ossec/framework/python/lib/python3.7/site-packages/Werkzeug-0.14.1-py3.7.egg/werkzeug/datastructures.py:16: DeprecationWarning: Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated, and in 3.8 it will stop working
    from collections import Container, Iterable, MutableSet

-- Docs: https://docs.pytest.org/en/latest/warnings.html
======================================== 4 passed, 4 warnings in 1.34 seconds ========================================

Regards

### CODE REVIEW

**Comment by crd1985** [comment]
I have integrated the functionality with openapi 3 spec. 
Only one endpoint allow basic authentication method through user/password and this must be used to get a valid token in a jwt format.
The rest of endpoint must use this valid token in order to been granted access.

The first time authentication module is imported two very important things happen: a new users.db database is initialized and a random secret is generated under wazuh_home_path/api.
users.db is a sqlite database storing user-pass pairs. Passwords are safely hash encoded. There are two default users that are created and cannot be deleted, wazuh and wazuh-app, with both passwords wazuh and wazuh-app. These passwords might be changed in a short future feature, although the logic for it is implemented in this PR.
The random secret is stored in a file under wazuh_home_path/api and must be readable only for root or wazuh user.

These are some examples of how is working:
```bash
root@03c5a6c3a9cf:/# curl -XGET "http://localhost:8080/agents"
{
  "detail": "No authorization token provided",
  "status": 401,
  "title": "Unauthorized",
  "type": "about:blank"
}
root@03c5a6c3a9cf:/# curl -u wazuh:wazuh -XGET "http://localhost:8080/security/user/authenticate"
{
  "token": "eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJ3YXp1aCIsImlhdCI6MTU1MTc5NjgwOCwiZXhwIjoxNTUxNzk3NDA4LCJzdWIiOiJmb28ifQ.gqDO-YVcZw8veTmqyPJgTPuR0BDGZLvmUVpj8ojjkzY"
}
root@03c5a6c3a9cf:/# curl -H  "Authorization: Bearer eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJ3YXp1aCIsImlhdCI6MTU1MTc5NjgwOCwiZXhwIjoxNTUxNzk3NDA4LCJzdWIiOiJmb28ifQ.gqDO-YVcZw8veTmqyPJgTPuR0BDGZLvmUVpj8ojjkzY" -XGET "http://localhost:8080/agents"
{
  "data": {
    "items": [
      {
        "dateAdd": "2019-02-21 14:49:28",
        "id": "000",
        "ip": "127.0.0.1",
        "lastKeepAlive": "9999-12-31 23:59:59",
        "manager": "03c5a6c3a9cf",
        "name": "03c5a6c3a9cf",
        "node_name": "master-node",
        "os": {
          "arch": "x86_64",
          "codename": "Bionic Beaver",
          "major": "18",
          "minor": "04",
          "name": "Ubuntu",
          "platform": "ubuntu",
          "uname": "Linux |03c5a6c3a9cf |4.15.0-43-generic |#46-Ubuntu SMP Thu Dec 6 14:45:28 UTC 2018 |x86_64",
          "version": "18.04.1 LTS"
        },
        "status": "Active",
        "version": "Wazuh v3.9.0"
      },
      {
        "configSum": "ab73af41699f13fdd81903b5f23d8d00",
        "dateAdd": "2019-03-05 12:05:40",
        "group": [
          "default"
        ],
        "id": "001",
        "ip": "172.18.0.5",
        "lastKeepAlive": "2019-03-05 14:42:34",
        "manager": "03c5a6c3a9cf",
        "mergedSum": "f8d49771911ed9d5c45b03a40babd065",
        "name": "1ecca129b329",
        "node_name": "master-node",
        "os": {
          "arch": "x86_64",
          "codename": "Bionic Beaver",
          "major": "18",
          "minor": "04",
          "name": "Ubuntu",
          "platform": "ubuntu",
          "uname": "Linux |1ecca129b329 |4.15.0-43-generic |#46-Ubuntu SMP Thu Dec 6 14:45:28 UTC 2018 |x86_64",
          "version": "18.04.1 LTS"
        },
        "registerIP": "172.18.0.5",
        "status": "Active",
        "version": "Wazuh v3.8.2"
      },
      {
        "configSum": "ab73af41699f13fdd81903b5f23d8d00",
        "dateAdd": "2019-03-05 12:05:45",
        "group": [
          "default"
        ],
        "id": "002",
        "ip": "172.18.0.6",
        "lastKeepAlive": "2019-03-05 14:42:25",
        "manager": "3f889099f363",
        "mergedSum": "f8d49771911ed9d5c45b03a40babd065",
        "name": "bc8af3e88e6c",
        "node_name": "worker1",
        "os": {
          "arch": "x86_64",
          "codename": "Bionic Beaver",
          "major": "18",
          "minor": "04",
          "name": "Ubuntu",
          "platform": "ubuntu",
          "uname": "Linux |bc8af3e88e6c |4.15.0-43-generic |#46-Ubuntu SMP Thu Dec 6 14:45:28 UTC 2018 |x86_64",
          "version": "18.04.1 LTS"
        },
        "registerIP": "172.18.0.6",
        "status": "Active",
        "version": "Wazuh v3.8.2"
      },
      {
        "configSum": "ab73af41699f13fdd81903b5f23d8d00",
        "dateAdd": "2019-03-05 12:05:46",
        "group": [
          "default"
        ],
        "id": "003",
        "ip": "172.18.0.7",
        "lastKeepAlive": "2019-03-05 14:42:28",
        "manager": "f7b349d65e04",
        "mergedSum": "f8d49771911ed9d5c45b03a40babd065",
        "name": "2d7dd6b566b5",
        "node_name": "worker2",
        "os": {
          "arch": "x86_64",
          "codename": "Bionic Beaver",
          "major": "18",
          "minor": "04",
          "name": "Ubuntu",
          "platform": "ubuntu",
          "uname": "Linux |2d7dd6b566b5 |4.15.0-43-generic |#46-Ubuntu SMP Thu Dec 6 14:45:28 UTC 2018 |x86_64",
          "version": "18.04.1 LTS"
        },
        "registerIP": "172.18.0.7",
        "status": "Active",
        "version": "Wazuh v3.8.2"
      }
    ],
    "totalItems": 4
  },
  "error": 0
}
root@03c5a6c3a9cf:/# curl -H  "Authorization: Bearer badtoken" -XGET "http://localhost:8080/agents"
{
  "detail": "The server could not verify that you are authorized to access the URL requested.  You either supplied the wrong credentials (e.g. a bad password), or your browser doesn't understand how to supply the credentials required.",
  "status": 401,
  "title": "Unauthorized",
  "type": "about:blank"
}
root@03c5a6c3a9cf:/# curl -u wazuh:wazuh -XGET "http://localhost:8080/agents"
{
  "detail": "No authorization token provided",
  "status": 401,
  "title": "Unauthorized",
  "type": "about:blank"
}
```

File: api/scripts/wazuh-apid.py [hunk 1]
```diff
@@ -75,6 +75,10 @@ def print_version():
 
 def main(cors, port, host, ssl_context):
     app = connexion.App(__name__, specification_dir=f'{common.ossec_path}/api/api/spec/')
+    for h in logger.handlers:
```
Comment (mgmacias95) [inline api/scripts/wazuh-apid.py]: These handlers are going to be added in #2739

File: api/setup.py [hunk 2]
```diff
@@ -19,7 +19,9 @@
 
 REQUIRES = ["connexion[swagger-ui]==2.2.0",
             "python_dateutil==2.6.0",
-            "setuptools>=21.0.0"
+            "setuptools>=21.0.0",
```
Comment (mgmacias95) [inline api/setup.py]: Please include these dependencies in the requirements.txt file too

Comment (crd1985) [reply api/setup.py]: I have redesigned the requirements.txt file to include only wazuh, api and dependencies from module.

File: api/scripts/wazuh-apid.py (hunk 1 above)
Comment (crd1985) [reply api/scripts/wazuh-apid.py]: Fixed!

**Pull request merged**